### ISSUE TITLE
TLS config is not detected resulting in error

### ISSUE THREAD

**Issue by Legion2** [testcontainers/testcontainers-node#315]
We run testcontainers in GitLab CI and configured docker dind with TLS verification which is enabled by default in docker since docker 19.03.12. We have setup the runner correctly, so the tls certificats are available and environment variables are set. `docker version` in the CI works correctly and can connect to dind via tls connection. However, testcontainer fails to run the tests with the following Debug log:

```
2021-12-22T18:43:24.335Z testcontainers DEBUG Found applicable Docker client strategy: ConfigurationStrategy
2021-12-22T18:43:24.336Z testcontainers DEBUG Testing Docker client strategy URI: tcp://docker:2376
2021-12-22T18:43:24.423Z testcontainers WARN  Docker daemon is not reachable: Error: error:0480006C:PEM routines::no start line
2021-12-22T18:43:24.424Z testcontainers WARN  Docker client strategy ConfigurationStrategy is not reachable
2021-12-22T18:43:24.424Z testcontainers DEBUG Found applicable Docker client strategy: UnixSocketStrategy
2021-12-22T18:43:24.425Z testcontainers DEBUG Testing Docker client strategy URI: unix:///var/run/docker.sock
```

I seems testcontainer tries to connect to the docker daemon without tls using the tls port.

**Comment by cristianrgreco** [comment]
hi @Legion2, how is the TLS configuration being provided to testcontainers? Is it via the environment or a `.testcontainers.properties` file? Could you please share the value of that configuration?

**Comment by Legion2** [comment]
It is provided as environment variables, they are set by the ci template:
```
    DOCKER_HOST: "tcp://docker:2376"
    DOCKER_DRIVER: overlay2
    DOCKER_TLS_CERTDIR: "/certs"
    DOCKER_TLS_VERIFY: 1
    DOCKER_CERT_PATH: "$DOCKER_TLS_CERTDIR/client"
```

**Comment by geordiepaul8** [comment]
Hi, is there an update on this. I am noticing the same issue. It works using v 8.0.0, but when I go to a higher version I keep getting: 

`2022-02-07T14:11:47.681Z testcontainers WARN  Docker client strategy ConfigurationStrategy is not reachable
2022-02-07T14:11:47.682Z testcontainers DEBUG Found applicable Docker client strategy: UnixSocketStrategy
2022-02-07T14:11:47.682Z testcontainers DEBUG Testing Docker client strategy URI: unix:///var/run/docker.sock
2022-02-07T14:11:47.702Z testcontainers WARN  Docker daemon is not reachable: Error: connect ECONNREFUSED 127.0.0.1:2375
2022-02-07T14:11:47.702Z testcontainers WARN  Docker client strategy UnixSocketStrategy is not reachable`

which leads to the error: 

`2022-02-07T14:12:01.625Z testcontainers INFO  Starting DockerCompose environment: testcontainers-647d4e229d967ab8cf243476728b8a76
2022-02-07T14:12:01.626Z testcontainers DEBUG Creating new Reaper for session: 33aa526b0119e27ee333aa27bcd099c4
2022-02-07T14:12:01.627Z testcontainers DEBUG No registry auth locator found for registry: "https://index.docker.io/v1/"
2022-02-07T14:12:01.628Z testcontainers ERROR Failed to list images: Error: No Docker client strategy found
2022-02-07T14:12:01.628Z testcontainers ERROR Failed to pull image "testcontainers/ryuk:0.3.2": Error: No Docker client strategy found`

all using gitlab-ci

**Comment by cristianrgreco** [comment]
Apologies @Legion2 and @geordiepaul8, I am swamped with work and other projects at the mo, any help with a PR would be much appreciated!

**Comment by geordiepaul8** [comment]
Hi @cristianrgreco just to update you on this, I got my issue working. With v8.4.0 I could not get the testcontainers to reach the docker deamon on gitlab CI using the docker-in-docker runner.

I used the following settings to get it to work:

```
TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE: /var/run/docker.sock
TESTCONTAINERS_HOST_OVERRIDE: docker
```
* note I did not set the `DOCKER_HOST` variable as the `tcp://` could not connect

I realise that this may not work for everyone, but if people have come across the same issue, maybe they can use this config and try it out.

**Comment by ferntheplant** [comment]
@Legion2 did you ever resolve this? I have a similar issue with my GitLab runner (which given the env variables I think you did too)

**Comment by Legion2** [comment]
@Legion2 no I still use the non tls url for node testcontainers. With java testcontainers library the tls configuration works.

**Comment by cristianrgreco** [comment]
@ferntheplant, @Legion2 could you please share how I can setup an environment to reproduce this? i should be able to have a look soon

**Comment by Legion2** [comment]
@cristianrgreco Run the `docker:dind` container such that it is accessible via `docker` hostname (docker compose for example)
Add a volume called `docker-certs` which is mounted in both docker dind and test container at `/certs/client`, docker dind will store the tls cert in there and the test container should read and use these.
Set the following environment variables in the test container
```
    DOCKER_HOST: "tcp://docker:2376"
    DOCKER_DRIVER: overlay2
    DOCKER_TLS_CERTDIR: "/certs"
    DOCKER_TLS_VERIFY: 1
    DOCKER_CERT_PATH: "$DOCKER_TLS_CERTDIR/client"
```

**Comment by ferntheplant** [comment]
I actually reverted to 8.0.0 and TLS now it works great.

My setup on 8.4.0 that failed is:
```typescript
process.env('DEBUG') // testcontainers*

const container = new GenericContainer(`docker`)
  .withNetworkMode('some-network')
  .withEnv('DOCKER_HOST', 'tcp://docker:2376')
  .withEnv('DOCKER_TLS_CERTDIR', '/certs')
  .withEnv('DOCKER_TLS_VERIFY', 1)
  .withEnv('DOCKER_CERT_PATH', '/certs/client')
  .withBindMount('some-docker-certs-client', '/certs/client')

const startedContainer = await container.start()

const { output } = startedContainer.exec(['docker', 'run', 'hello-world'])  // hangs
```

```shell
$ docker run --privileged --name some-docker -d \
    --network some-network --network-alias docker \
    -e DOCKER_TLS_CERTDIR=/certs \
    -v some-docker-certs-ca:/certs/ca \
    -v some-docker-certs-client:/certs/client \
    docker:dind
```
I get the same testcontainers log output as Legion in the issue description, specifically `Docker daemon is not reachable: Error: error:0480006C:PEM routines::no start line` causing the testcontainer module to default to the host docker engine.

**Comment by cristianrgreco** [comment]
@ferntheplant I need some more help reproducing this. This is the test I've written:

```typescript
it("should connect to the Docker daemon over TLS", async () => {
  const certsDir = path.resolve(__dirname, "..", "..", "fixtures", "certs");

  const network = await new Network().start();

  const dockerTls = await new GenericContainer("docker:dind")
    .withPrivilegedMode()
    .withExposedPorts(2376)
    .withNetworkMode(network.getName())
    .withNetworkAliases("docker-tls")
    .withBindMount(certsDir, "/certs")
    .withEnv("DOCKER_TLS_CERTDIR", "/certs")
    .start();

  const container = await new GenericContainer("docker")
    .withCmd(["sleep", "infinity"])
    .withNetworkMode(network.getName())
    .withEnv("DOCKER_HOST", `tcp://docker-tls:${dockerTls.getMappedPort(2376)}`)
    .withEnv("DOCKER_TLS_VERIFY", "1")
    .withEnv("DOCKER_CERT_PATH", "/certs")
    .withBindMount(certsDir, "/certs")
    .start();

  const { output } = await container.exec(["docker", "run", "hello-world"]);
  expect(output).toContain("Hello from Docker!");

  await container.stop();
  await dockerTls.stop();
  await network.stop();
});
```

I've generated the certs according to: https://docs.docker.com/engine/security/protect-access/#create-a-ca-server-and-client-keys-with-openssl, specifically:

```
➜  docker-certs openssl genrsa -aes256 -out ca-key.pem 4096
Generating RSA private key, 4096 bit long modulus
...................................................................................++
.......................................................................................................++
e is 65537 (0x10001)
Enter pass phrase for ca-key.pem:
Verifying - Enter pass phrase for ca-key.pem:

➜  docker-certs openssl req -new -x509 -days 365 -key ca-key.pem -sha256 -out ca.pem
Enter pass phrase for ca-key.pem:
You are about to be asked to enter information that will be incorporated
into your certificate request.
What you are about to enter is what is called a Distinguished Name or a DN.
There are quite a few fields but you can leave some blank
For some fields there will be a default value,
If you enter '.', the field will be left blank.
-----
Country Name (2 letter code) []:.
State or Province Name (full name) []:.
Locality Name (eg, city) []:.
Organization Name (eg, company) []:.
Organizational Unit Name (eg, section) []:.
Common Name (eg, fully qualified host name) []:docker-tls
Email Address []:.

➜  docker-certs openssl genrsa -out server-key.pem 4096
Generating RSA private key, 4096 bit long modulus
.................................++
.........................................................................++
e is 65537 (0x10001)

➜  docker-certs openssl req -subj "/CN=$HOST" -sha256 -new -key server-key.pem -out server.csr

➜  docker-certs echo subjectAltName = DNS:$HOST,IP:10.10.10.20,IP:127.0.0.1 >> extfile.cnf

➜  docker-certs echo extendedKeyUsage = serverAuth >> extfile.cnf

➜  docker-certs openssl x509 -req -days 365 -sha256 -in server.csr -CA ca.pem -CAkey ca-key.pem \
  -CAcreateserial -out server-cert.pem -extfile extfile.cnf
Signature ok
subject=/CN=docker-tls
Getting CA Private Key
Enter pass phrase for ca-key.pem:

➜  docker-certs openssl genrsa -out key.pem 4096
Generating RSA private key, 4096 bit long modulus
.........................................................................................................................................................................................................................................................................................................++
....................................++
e is 65537 (0x10001)

➜  docker-certs openssl req -subj '/CN=client' -new -key key.pem -out client.csr

➜  docker-certs echo extendedKeyUsage = clientAuth > extfile-client.cnf

➜  docker-certs openssl x509 -req -days 365 -sha256 -in client.csr -CA ca.pem -CAkey ca-key.pem \
  -CAcreateserial -out cert.pem -extfile extfile-client.cnf
Signature ok
subject=/CN=client
Getting CA Private Key
Enter pass phrase for ca-key.pem:

➜  docker-certs rm -v client.csr server.csr extfile.cnf extfile-client.cnf
client.csr
server.csr
extfile.cnf
extfile-client.cnf
```

Note how I've set the CN to `docker-tls`. However when I run the test, `docker:dind` starts up successfully, however `docker` emits the following error:

```
http: TLS handshake error from 127.0.0.1:38707: EOF 
```

I get this error regardless if I run testcontainers 8.4.0 or 8.0.0. Any help would be appreciated!

**Comment by Legion2** [comment]
The certificates are automatically generated by docker:dind see https://hub.docker.com/_/docker/#tls "Starting in 18.09+, the dind variants of this image will automatically generate TLS certificates in the directory specified by the DOCKER_TLS_CERTDIR environment variable."

**Comment by cristianrgreco** [comment]
Thanks @Legion2, I'm now letting dind create the certs, and i'm mounting the client certs from that volume into the docker container. Here's the updated test:

```typescript
it("should connect to the Docker daemon over TLS", async () => {
  const certsDir = path.resolve(__dirname, "..", "..", "fixtures", "certs");

  const network = await new Network().start();

  const dockerTls = await new GenericContainer("docker:dind")
      .withPrivilegedMode()
      .withExposedPorts(2376)
      .withNetworkMode(network.getName())
      .withNetworkAliases("docker-tls")
      .withBindMount(certsDir, "/certs")
      .withEnv("DOCKER_TLS_CERTDIR", "/certs")
      .start();

  const container = await new GenericContainer("docker")
      .withCmd(["sleep", "infinity"])
      .withNetworkMode(network.getName())
      .withEnv("DOCKER_HOST", `tcp://docker-tls:2376`)
      .withEnv("DOCKER_TLS_VERIFY", "1")
      .withEnv("DOCKER_CERT_PATH", "/certs/client")
      .withBindMount(certsDir, "/certs")
      .start();

  const { output } = await container.exec(["docker", "run", "hello-world"]);
  expect(output).toContain("Hello from Docker!");

  await container.stop();
  await dockerTls.stop();
  await network.stop();
});
```

Next problem is this (also same from 8.0.0/latest): 
```
http: TLS handshake error from 172.19.0.3:46188: remote error: tls: bad certificate
```

**Comment by cristianrgreco** [comment]
@Legion2 I think I've found the issue, could you please try version `8.6.1-beta.0` and let me know?

**Comment by Legion2** [comment]
@cristianrgreco thanks with `8.6.1-beta.0` the our tests work with tls enabled

**Comment by cristianrgreco** [comment]
Excellent, released in 8.6.1. Thanks for your help!
